compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -12810,6 +12810,149 @@ TR::Node *su2dSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s)
+*/
+TR::Node* removeArithmeticsUnderIntegralCompare(TR::Node* node,
+                                                TR::Simplifier * s)
+   {
+   if (s->comp()->getOption(TR_DisableIntegerCompareSimplification))
+      return node;
+
+   TR::Node* opNode      = node->getFirstChild();
+   TR::Node* secondChild = node->getSecondChild();
+   TR::ILOpCodes op      = opNode->getOpCodeValue();
+   bool isInt            = !opNode->getOpCode().isLong();
+
+   // Get recognized operation types. Unsigned IL OpCodes are deprecated and not supported here.
+   bool isAddOp = (op == TR::iadd || op == TR::ladd || op == TR::sadd || op == TR::badd);
+   bool isSubOp = (op == TR::isub || op == TR::lsub || op == TR::ssub || op == TR::bsub);
+
+   if ((isAddOp || isSubOp) &&
+           opNode->chkCannotOverflow()  &&
+           opNode->getSecondChild()->getOpCode().isLoadConst() &&
+           secondChild->getOpCode().isLoadConst() &&
+           // Only perform the simplification on the first occurance of the add/sub node
+           (opNode->getFutureUseCount() == opNode->getReferenceCount() - 1))
+      {
+      bool isUnsignedCompare = node->getOpCode().isUnsignedCompare();
+      int64_t oldConst1 = 0, oldConst2 = 0, newConst = 0;   // for all transformations other than unsigned long
+      uint64_t newUConst = 0;                               // for unsigned long comparisons
+
+      // Get and check the integer/long constant values
+      if (isInt)
+         {
+         oldConst1 = opNode->getSecondChild()->getConstValue();
+         oldConst2 = secondChild->getConstValue();
+         newConst = isAddOp ? (oldConst2 - oldConst1) : (oldConst2 + oldConst1);
+
+         // Get type ranges to check for overflow or underflow
+         int64_t signedMax = 0, signedMin = 0, unsignedMax = 0, unsignedMin = 0;
+         if (opNode->getOpCode().is1Byte())
+            {
+            signedMax   = TR::getMaxSigned<TR::Int8>();
+            signedMin   = TR::getMinSigned<TR::Int8>();
+            unsignedMax = TR::getMaxUnsigned<TR::Int8>();
+            unsignedMin = TR::getMinUnsigned<TR::Int8>();
+            }
+         else if (opNode->getOpCode().is2Byte())
+            {
+            signedMax   = TR::getMaxSigned<TR::Int16>();
+            signedMin   = TR::getMinSigned<TR::Int16>();
+            unsignedMax = TR::getMaxUnsigned<TR::Int16>();
+            unsignedMin = TR::getMinUnsigned<TR::Int16>();
+            }
+         else if (opNode->getOpCode().is4Byte())
+            {
+            signedMax   = TR::getMaxSigned<TR::Int32>();
+            signedMin   = TR::getMinSigned<TR::Int32>();
+            unsignedMax = TR::getMaxUnsigned<TR::Int32>();
+            unsignedMin = TR::getMinUnsigned<TR::Int32>();
+            }
+         else
+            return node;
+
+         if ((!isUnsignedCompare && (newConst > signedMax || newConst < signedMin)) ||
+                 (isUnsignedCompare && (newConst > unsignedMax || newConst < unsignedMin)))
```

Comment:
indent here seems large

Reply:
This overflow test is sensitive to whether the old constants were set using `setInt()` or `setUnsignedInt()` (or similar for narrower types). For example, suppose we have a signed 32-bit comparison `x + 1 <= INT_MIN`. If this `INT_MIN` comes from `setInt()`, then overflow would be correctly detected. OTOH, if it comes from `setUnsignedInt()`, then I believe we'd get `newConst == (int64_t)INT_MAX`, which would incorrectly appear not to have overflowed